### PR TITLE
feature: added transfer ERC-20 tokens

### PR DESCRIPTION
**Pull request by piotrwitek**
## Description

implemented happy path for complete transfer flow with ui forms, updated UI controls to support erc-20 tokens and added new validation for erc-20 tokens forms

## Linked tickets

(https://make-software.atlassian.net/browse/WALLET-74)

## Checklist

- [x] Make sure this PR title follows semantic release conventions: <https://semantic-release.gitbook.io/semantic-release/#commit-message-format>

- [x] If the PR adds any new text to the UI, make sure they are localized

- [ ] Include a screenshot or recording if implementing significant UI or user flow change

- [ ] When this PR affects architecture changes wait for review from Piotr before merging

### CODE REVIEW

**Comment by piotrwitek** [comment]
@ost-ptk It's WIP but good if you can have a look early :)

**Comment by piotrwitek** [comment]
Hey @ost-ptk could you please have a look?

**Comment by ost-ptk** [comment]
Hey @piotrwitek 
I can't run the project on this branch. There are a few errors in the console.

**Comment by piotrwitek** [comment]
That's Casper event parser dependency

File: src/apps/popup/index.html [hunk 1]
```diff
@@ -2,11 +2,7 @@
 <html lang="en">
   <head>
     <meta charset="UTF-8" />
-    <meta
```
[the remark this answers is not in the document]
Comment (piotrwitek) [reply src/apps/popup/index.html:5]: For local testing it should be rolled back

File: src/apps/popup/pages/transfer/amount-step.tsx [hunk 2]
```diff
@@ -85,6 +96,37 @@ export const AmountStep = ({ amountForm }: AmountStepProps) => {
...
+        <TransferInputContainer>
+          <Input
+            label={paymentAmoutLabel}
+            rightLabel={paymentFiatAmount}
+            type="number"
+            monotype
+            placeholder={t('Enter transaction fee')}
+            suffixText={'CSPR'}
+            {...register('paymentAmount')}
+            error={!!errors?.transferIdMemo}
```
[the remark this answers is not in the document]
Comment (piotrwitek) [reply src/apps/popup/pages/transfer/amount-step.tsx:122]: Good catch

[the remark this answers is not in the document]
Comment (piotrwitek) [reply src/apps/popup/pages/transfer/amount-step.tsx:122]: Fixed

File: src/apps/popup/pages/transfer/amount-step.tsx [hunk 3]
```diff
@@ -85,6 +96,37 @@ export const AmountStep = ({ amountForm }: AmountStepProps) => {
           validationText={errors?.transferIdMemo?.message}
         />
       </TransferInputContainer>
+      {/** transferIdMemo is only relevant for CSPR */}
+      {(isCSPR && (
+        <TransferInputContainer>
```
[the remark this answers is not in the document]
Comment (piotrwitek) [reply src/apps/popup/pages/transfer/amount-step.tsx:101]: Removed duplication, thanks.

File: src/apps/popup/index.html (hunk 1 above)
[the remark this answers is not in the document]
Comment (piotrwitek) [reply src/apps/popup/index.html:5]: Fixed

**Comment by piotrwitek** [comment]
@ost-ptk please re-review and let me know if looks good, thanks

File: src/apps/popup/app-router.tsx [hunk 4]
```diff
@@ -256,7 +256,7 @@ function AppRoutes() {
         path={RouterPath.DownloadedSecretKeys}
         element={<DownloadedSecretKeysPage />}
       />
-      <Route path={RouterPath.Transfer} element={<TransferPage />} />
+      <Route path={RouterPath.TransferERC20} element={<TransferPage />} />
```
Comment (ost-ptk) [inline src/apps/popup/app-router.tsx:259]: I guess we need to rename this route path. It is responsible for native transfer too

**Comment by ost-ptk** [comment]
Also, I noticed that we make a call for a contract package endpoint every time when going to the home page. I guess we need to add some timeout as we add it to all other request

**Comment by piotrwitek** [comment]
> Also, I noticed that we make a call for a contract package endpoint every time when going to the home page. I guess we need to add some timeout as we add it to all other request

requests are cached by react-query

File: src/apps/popup/app-router.tsx (hunk 4 above)
Comment (piotrwitek) [reply src/apps/popup/app-router.tsx:259]: how do you want to rename it?

**Comment by piotrwitek** [comment]
@ost-ptk I have updated `@make-software/ces-js-parser` dependency and resolved all conflicts with the base branch, please have a look

File: src/apps/popup/app-router.tsx (hunk 4 above)
Comment (ost-ptk) [reply src/apps/popup/app-router.tsx:260]: I think `TokenTransfer` or `Transfer`.
Now this route is responsible for erc20 and native transfer, so it's confusing when it names `TransferERC20`

**Comment by ost-ptk** [comment]
> > Also, I noticed that we make a call for a contract package endpoint every time when going to the home page. I guess we need to add some timeout as we add it to all other request
> 
> requests are cached by react-query, but feel free to suggest an alternative solution

Than something goes wrong. Please take a look

https://github.com/make-software/casper-wallet/assets/44294945/8076c8e5-cb6a-4ff7-98b5-044f151bb041

**Comment by ost-ptk** [comment]
> > We need to add some loader/skeleton or something to erc20 token detail page to prevent blinking UI.
> 
> Is the UI blinking somewhere? could you record it, please? I haven't noticed.

Sure, please take a look

https://github.com/make-software/casper-wallet/assets/44294945/ff1c249e-57f7-4793-9dd5-b54e22f1ded2

File: src/apps/popup/pages/receive/content.tsx [hunk 5]
```diff
@@ -43,7 +43,7 @@ export const ReceivePageContent = () => {
           <Trans t={t}>Receive tokens</Trans>
         </Typography>
       </ParagraphContainer>
-      <ActiveAccountPlate label="To account" />
+      <ActiveAccountPlate label="To account" symbol={'CSPR'} balance={null} />
```
Comment (ost-ptk) [inline src/apps/popup/pages/receive/content.tsx:46]: Why you didn't pass the balance prop here?

Comment (piotrwitek) [reply src/apps/popup/pages/receive/content.tsx:46]: This is not related to this work

**Comment by ost-ptk** [comment]
I tried to send a token and received an error
Знімок екрана 2023-06-28 о 11 19 11

**Comment by piotrwitek** [comment]
@ost-ptk what are STR? could you please record a video for the bug so I can fix it?

**Comment by piotrwitek** [comment]
> > > Also, I noticed that we make a call for a contract package endpoint every time when going to the home page. I guess we need to add some timeout as we add it to all other request
> > 
> > 
> > requests are cached by react-query, but feel free to suggest an alternative solution
> 
> Than something goes wrong. Please take a look
> 
>  2023-06-28.10.46.57.mov

There was a bug in the fetchContractPackage function I fixed it

**Comment by ost-ptk** [comment]
> @ost-ptk what are STR? could you please record a video with the error above so I can fix it?

Sure. I record a video

https://github.com/make-software/casper-wallet/assets/44294945/8e38a01e-6274-4b9e-a5a1-68283944dd45

https://github.com/make-software/casper-wallet/assets/44294945/2e480738-9d55-4293-91cf-e2b7e555547f

**Comment by piotrwitek** [comment]
@ost-ptk I cannot reproduce your error reported in the PR, could you test the most recent commit and use the production build?

**Comment by ost-ptk** [comment]
I still have this error. We can try to merge this PR and run CI tests. Maybe this is my local issue
@piotrwitek What do you think?

**Comment by piotrwitek** [comment]
@ost-ptk agreed I'll merge and I can create a test build

**Pull request merged**